Updated query method for v0.13 and later server.(#329)

wx00 opened the pull request:
`query` now uses `GET` for  `SELECT` or `SHOW` queries, and `POST` for others.

babilen commented:
This would fix https://github.com/influxdata/influxdb-python/issues/329

wx00 commented:
Thanks.
I found it brokes test cases, and I'm now trying to update those test queries that are not `SELECT` or `SHOW` with `GET` method.
May update tonight or tomorrow (using China Standard Time).

babilen commented:
Which test cases are these?

wx00 commented:
I think it's ALL queries tests except those using `SELECT` and `SHOW` queries. So, a lot of the tests should be modified...

babilen commented:
I didn't see a difference in tests after cherry picking your commit. It's still a needed change though :)

wx00 commented:
I found the problem after sending this PR, didn't make changes in tests yet, will do it later in a few days..

wx00 commented:
@babilen @aviau, can you help me with the test cases? Looks like the influxdb v0.12.1 doesn't support `POST` method in some queries.

babilen commented:
Please rebase your branch against master and force push to GH. That will include a number of fixes that are necessary to get this running. In particular it will ensure that InfluxDB 0.13 is installed.

wx00 commented:
I modified client_test.py and it passed all tests in py2.7 on my test machine now. Do I need to send another pull request after update?

wx00 commented:
conflicts [here?](https://github.com/influxdata/influxdb-python/compare/master...wx00:master#diff-58c51d2852377954867cd17a8b327af0R718)

babilen commented:
The problem is that you developed in your "master" branch and didn't rebase when you pulled in changes. I would generally recommend to create a specific feature branch (e.g. `git checkout -b query-method`), pull upstream changes into master (`git pull` in the master branch) and rebase your branch against master every now and then (`git rebase master` in your `query-method` branch).

Your master branch is in some strange state right now and would probably have to be reset. I checked your commit and the tests pass and it applies cleanly though, so the question is how to clean up this PR.

Maybe a `git pull --rebase` is all you need in your master branch, but maybe you have to create the feature branch, hard reset your master branch to some earlier commit and rebase your feature branch against master with `git rebase master`.

babilen commented:
Would `query.split()[0].upper()` work with an empty query? It looks as if that would throw an `IndexError` that doesn't make it easy for the user to understand what went wrong.

It would be nice if you could default to one method and then override it *if* a query has been passed and if it starts with either `SELECT` or `SHOW`. We could also think about using `.startswith(...)` here.

wx00 commented:
Good point, I didn't think of empty string, that would throw an `IndexError` as you said.
The reason I use `split()` here is, if the query is start with whitespace character, it can still get the first word.
It looks like `.startswith(...)` can only match one word at a call? 

How do you think about this code?
``` python
method = 'POST'
if query.split() and query.split()[0].upper() in ('SELECT', 'SHOW'):
    method = 'GET'
```

wx00 commented:
I'm not so familiar with all those git branch operations, I'll try to fix it, thanks for the help.

babilen commented:
Wonderful, thank you for looking into this. `.startswith(...)` would only be able to look at one string at a time, yeah.

I personally wouldn't mind:

```
method = 'POST'
normalised_query = query.lower()
if normalised_query:
  if query.startswith('select'):
    method = 'GET'
  elif query.startswith('show'):
    method = 'GET'
```

Or parsing the query explicitly like you do. I simply don't like chaining multiple method calls and index lookups in the same statement (things like `split()[0].upper()`) and tend to split it into multiple steps.

@aviau What do you think would be a good approach here?

wx00 commented:
Hmm..
 I hard reset `master` and pulled from upstream, and it's synced with upstream then(check by `git log`), then I apply a patch from my previous "future" commit. So the codes should be fine and now with a clean commit history.
But I just don't sure how it will behave when I push onto github... Is it OK to push?

wx00 commented:
Your method here still has the problem when the query starts with whitespace characters and then `select` or `show`..
No problem if it's no characters in string, it will be dealt in the later `request(...)`.

babilen commented:
You can't break anything by force pushing into your own repository and it sounds as if you have a clean branch now.

I agree that the approach I sketched above does have issues too. I simply wanted to sketch the idea, rather than provide a full implementation. Basically just wanted to ensure that the data handling is robust.

rezen commented:
Can we get this merged in?

wx00 commented:
Outdated
